### DISCUSSION TITLE
issue: Select with multiple attribute not resetting

### DISCUSSION THREAD

**Discussion by giorgiabosello** [react-hook-form/react-hook-form#8790] [General] [answered]
### Version Number

7.34.0

### Codesandbox/Expo snack

https://codesandbox.io/s/react-hook-form-select-multiple-2250cm?file=/src/App.js

### Steps to reproduce

1. Create multiple fields inside a form
2. Fields have nested name - `FIELD_NAME._{id}`
3. Use `resetField` to empty parent name -> `reset(FIELD_NAME)`
4. See error -> field with simple select and string value is updated, while multiple select and array value is not.

### Expected behaviour

Select and Multi select should be reset to form default values when using `resetField` on parent name, instead, only default select will reset.
Value of the select isn't updated when it is an array type (multi select), while string is. 
`getValues` of `useFormContext` is correct though.

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by Moshyfawn** [comment]
You should consider
- **either** registering the whole `FIELD_NAME` field and updating partial field value like `onChange({ ...value, _1: event.target.value })`
- **or** resetting each specific field individually

**Comment by giorgiabosello** [comment]
> You should consider
> 
> * **either** registering the whole `FIELD_NAME` field and updating partial field value like `onChange({ ...value, _1: event.target.value })`
> * **or** resetting each specific field individually

I am resetting the whole `FIELD_NAME`, but it doesn't update the multi select value. `onChange` isn't triggered.
`value` in `useController` is still the old one, while `getValues(FIELD_NAME)` is correct.

**Comment by bluebill1049** [comment]
https://react-hook-form.com/api/useform/resetfield

> Reset an individual field state

`resetField` API is made to reset a single field for multiple inputs and use `reset` instead or you can do the following

```
resetField(fieldname1')
resetField(fieldname2');
```

**Reply by giorgiabosello** [reply]
I'm using `resetField` as intended.
My default values are as this:
```
{
    ...other fields,
    [FIELD_NAME]: {
        _1: '',
        _2: [],
    },
    ...other fields,
}
```
My selects are, then: `FIELD_NAME._1` and `FIELD_NAME._2`.
Resetting `FIELD_NAME`, will reset to default `_1`, but not `_2`, and as I said earlier (and you can confirm by watching di RHF DevTool), on RHF values are back to default, it's just that _2 onChange with array, isn't triggered on the component, but if you log `getValues(FIELD_NAME._2)`, value is `[]`.

I cannot do a complete `reset` of the form, since I have other fields that I don't want to reset.

I think the problem is when resetting field that as a nested array value.

**Reply by bluebill1049** [reply]
The issue is the below, the `fieldName` is not the actual input name

```tsx
const handleReset = (fieldName) => {
  return resetField(fieldName);
};
```

instead of 

```tsx
name={`${FIELD_NAME}._1`}
``

could you try `resetField(`${FIELD_NAME}._1`)` instead?

**Reply by giorgiabosello** [reply]
Doing like you said, it works.
But why, doing `resetField(FIELD_NAME)`, `${FIELD_NAME}._1` is back to default, while `${FIELD_NAME}._2` isn't?

And, with ```getValues(`${FIELD_NAME}._2`)```, value is correct, while `const { value } = useController({ ... })` isn't?

### before reset
![before](https://user-images.githubusercontent.com/9432544/182812762-dc6d6bc1-e6b3-4df4-8900-9f245e2b3257.png)
### after reset
![after](https://user-images.githubusercontent.com/9432544/182812791-f69801d1-c3d7-4f8b-8ba8-c0bef298fcfa.png)

**Reply by bluebill1049** [reply]
> But why, doing resetField(FIELD_NAME), ${FIELD_NAME}._1 is back to default, while ${FIELD_NAME}._2 isn't?

because that is what the API was designed for `resetField` a single field, just as the doc described.

**Reply by giorgiabosello** [reply]
But it is a single field 😄 
So, how I should reset a single field that has nested values? Should I loop every nested value and reset them one by one? It's a little heavy, considering I couldn't know how many of them are and if they have other nested values.

**Reply by bluebill1049** [reply]
you are using `Controller` (in your example) twice each Controller represents a single field, I have suggested you reset API, or you will have to update the filled value with forEach with `resetField`.

**Reply by giorgiabosello** [reply]
Changing the value in ControlledSelect, with `value={props.multiple ? getValues(name) : value}` will work. 
So, what's wrong? 
If I need to update each filled value with `resetField` in a forEach, I would expect `_1` to not reset also, instead, `_1` is reset with `resetField(FIELD_NAME)`. 
One way or the other, something is wrong with the API: 
1. `FIELD_NAME._1` shouldn't be reset as `FIELD_NAME._2` OR
2. `FIELD_NAME._2` should be reset as `FIELD_NAME._1` ?

**Comment by giorgiabosello** [comment]
Any update about this issue?

Relative to the last comment, one way or the other, something is wrong with the API:

FIELD_NAME._1 shouldn't be reset as FIELD_NAME._2 OR
FIELD_NAME._2 should be reset as FIELD_NAME._1 ?

Thank you.

**Comment by leapful** [comment]
@giorgiabosello : There is nothing wrong with API "resetField", the API expected to reset a single field which is primitive value while the "select._2" field had an array value therefore it cannot be reset since it will conflict with real "field array" handling of react-hook-form. You should reset "select._2" field individually instead of just the whole form values with "select".

**Comment by giorgiabosello** [comment]
@leapful: it's just `value` from `useController` that has the wrong value. Instead, `getValues` from `useFormContext` is perfectly reset. 
In fact, changing the value in ControlledSelect, with `value={props.multiple ? getValues(name) : value}` will work.
If I want to reset a single field, I would use `resetField`, but I'm not and I would expect `reset` to reset all the fields correctly and to update/trigger all the values in all the hooks/functions.

**Comment by leapful** [comment]
@giorgiabosello : If you want to reset all fields with your default values, just use "reset" API instead of "resetField".

Working demo: https://codesandbox.io/s/react-hook-form-select-multiple-forked-pwfmwl?file=/src/App.js

**Comment by giorgiabosello** [comment]
@leapful yes, it is working like this, you are resetting the whole form to default values, but I don't want that. I want to reset only a specific field with nested values, and nested values are being reset, except the array one. 
And it happens only in the value passed to useController. Like I said, getValues has the correct reset value and you can confirm it by looking at the logs.

**Comment by leapful** [comment, marked as answer]
@giorgiabosello : The API "resetField" is not built for field value that is not primitive, you are using wrong tool for the job. 

Even if "getValues" returns your expected values while "value" from "useController" is not, it doesn't mean there is a bug in API "resetField". 

Furthermore, you said that `"you are resetting the whole form to default values, but I don't want that"` and `" I want to reset only a specific field with nested values"` while API "reset" or "resetField" with correct params will do the same expected result.

For example:

```
reset({
  ...DEFAULT_VALUES, // or CURRENT_FORM_VALUES in case there is another field should keep value.
   select: {
     _1: '' // Empty or any value.
     _2: [] // Empty array or any array items.
   }
})
```

or

```
resetField('select._1');
resetField('select_.2');
```

If it cannot achieve your expectation then please explain more details about why you cannot use those APIs like the example above ? What is the difficulty in your use case that make you cannot do it ?

Beside that, react-hook-form API documentation for "resetField" might be updated to notify mis-using case like this issue.

**Comment by giorgiabosello** [comment]
@leapful: nothing prevents me on using the APIs as you suggested in the last comment. I opened the discussion because I was expecting something different, never red anything about primitive values in the documentation, so I didn't know values weren't treated all the same. I will implement the reset as you suggested and mark the question as answered. 
Thank you!
